department-of-veterans-affairs/caseflow
# MattT/APPEALS-15369: Save SSNs Returned From BGS to veterans Table

Resolves [APPEALS-15369](https://vajira.max.gov/browse/APPEALS-15369)

### Description
We have numerous veteran records in our [veterans](https://github.com/department-of-veterans-affairs/caseflow/blob/d9bd0c5fbe55533d7bf7528f79a8b87eb1fd2624/db/schema.rb#L1713) table that are missing social security number values. Whenever SSNs are requested from these veteran records, Caseflow is set up to query BGS for these values. The issue this PR aims to resolve is the lack of functionality to cache this value in our `veterans` table to allow for the following:
- Proper sorting of decision review queues by SSN values
- Reducing unnecessary BGS calls across the application

### Acceptance Criteria
- [ ] Whenever .ssn is called upon a Veteran model object, and a BGS call occurs due to a missing SSN value, the SSN from the BGS response is saved in our Caseflow database.
  - [ ] Specifically, the ssn column value for the veteran record is updated.

### Testing Plan
1. Open the Rails console
2. Obtain a veteran record, and save it to a variable
```ruby
veteran = Veteran.order('RANDOM()').first
```
3. Clear this veteran's SSN
```ruby
veteran.update!(ssn: nil)
```
4. Verify that the SSN has been cleared
```ruby
pp veteran
=> #<Veteran:0x000055e5b59b4240
 id: 1298,
 bgs_last_synced_at: nil,
 closest_regional_office: nil,
 created_at: Tue, 07 Feb 2023 19:06:06 UTC +00:00,
 date_of_death: nil,
 date_of_death_reported_at: nil,
 file_number: "300000102",
 first_name: "Carmen",
 last_name: "Klocko",
 middle_name: nil,
 name_suffix: "1298",
 participant_id: "700000102",
 ssn: nil, # Should be nil
 updated_at: Tue, 07 Feb 2023 19:06:06 UTC +00:00>
 ```
 5. Call `veteran.ssn`. Ensure that you receive an SSN back from "BGS" (our mock instance of it, at least).
 
 ```ruby
 [11] pry(main)> veteran.ssn
[2023-02-09 17:11:51 -0500]    (0.6ms)  BEGIN
[2023-02-09 17:11:51 -0500]   Veteran Update (2.1ms)  UPDATE "veterans" SET "bgs_last_synced_at" = $1, "ssn" = $2, "updated_at" = $3 WHERE "veterans"."id" = $4  [["bgs_last_synced_at", "2023-02-09 22:11:51.698813"], ["ssn", "698775905"], ["updated_at", "2023-02-09 22:11:51.701411"], ["id", 1702]]
[2023-02-09 17:11:51 -0500]    (2.0ms)  COMMIT
=> "698775905"
```

You should see some ActiveRecord logs show that `ssn` and `bgs_last_synced_at` have been updated on the veteran record.
 6. Check that the SSN you've received in step 5 has been saved to the veteran record:
 ```ruby
 veteran.reload.ssn
 => "698775905"
 ```
 7. Check also the `bgs_last_synced_at` value. It should correspond to the current time at which you ran step 5.
 ```ruby
 veteran.bgs_last_synced_at
 => Thu, 09 Feb 2023 22:11:51 UTC +00:00
 ```
